### PR TITLE
[TR] PIM-5139: fix grid association performance and sorting issues

### PR DESCRIPTION
**Pull request by jjanvier**
Time to load the association grid with customer's data (env prod, local machine):
- before: 12s
- now: <1s

| Q                 | A
| ----------------- | ---
| Specs             | Y
| Behats            | -
| Blue CI           | running
| Changelog updated |  Y
| Review and 2 GTM  | -

### CODE REVIEW

File: src/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/AssociatedProductHydrator.php [hunk 1]
```diff
@@ -90,56 +124,89 @@ public function hydrate($queryBuilder, array $options = [])
...
+        $in = ['_id' => ['$in' => $associatedIds]];
+        $rawQuery['$and'][] = $in;
+
+        $queryBuilder->setQueryArray($rawQuery);
+        $queryBuilder->limit($limit);
+        $queryBuilder->skip($skip);
+
+        $query = $queryBuilder->getQuery();
+        $results = $query->execute()->toArray();
+        array_walk($results, function(&$product) { $product['is_associated'] = true; });
```
Comment (sumbobyboys) [inline src/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/AssociatedProductHydrator.php]: !()[http://i.imgur.com/MHuW96t.gif?noredirect]

File: src/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/AssociatedProductHydrator.php [hunk 2]
```diff
@@ -52,11 +53,44 @@ public function hydrate($queryBuilder, array $options = [])
...
+                $skip
+            );
+
+            //second query: fetch only non associated for this page
+            $nonAssociatedProducts = [];
+            $nbAssociated  = count($associatedProducts);
+            if ($limit > $nbAssociated) {
+                $limit -= $nbAssociated;
+                $skip -= $nbTotalAssociated;
+                // $skip can be < 0  for example when you have 12 associated products, each page contains 10 items,
```
Comment (sumbobyboys) [inline src/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/AssociatedProductHydrator.php]: please remove the double space :trollface:

File: spec/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/AssociatedProductHydratorSpec.php [hunk 3]
```diff
@@ -237,5 +152,6 @@ function () use (&$productsCount) {
 
         $firstResult = $rows[0];
         $firstResult->shouldBeAnInstanceOf('\Oro\Bundle\DataGridBundle\Datasource\ResultRecord');
+
```
Comment (phaseinducer) [inline spec/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/AssociatedProductHydratorSpec.php]: no

Comment (sumbobyboys) [inline spec/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/AssociatedProductHydratorSpec.php]: could you explain this modification ?

File: src/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/AssociatedProductHydrator.php (hunk 1 above)
Comment (BitOne) [inline src/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/AssociatedProductHydrator.php]: Foreach please. More readable (and faster but we should'nt care ;) ).

File: src/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/AssociatedProductHydrator.php [hunk 4]
```diff
@@ -90,56 +124,89 @@ public function hydrate($queryBuilder, array $options = [])
...
+        $nin = ['_id' => ['$nin' => $associatedIds]];
+        $rawQuery['$and'][] = $nin;
+
+        $queryBuilder->setQueryArray($rawQuery);
+        $queryBuilder->limit($limit);
+        $queryBuilder->skip($skip);
+
+        $query = $queryBuilder->getQuery();
+        $results = $query->execute()->toArray();
+        array_walk($results, function(&$product) { $product['is_associated'] = false; });
```
Comment (BitOne) [inline src/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/AssociatedProductHydrator.php]: Foreach.

File: src/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/AssociatedProductHydrator.php [hunk 5]
```diff
@@ -66,7 +100,7 @@ public function hydrate($queryBuilder, array $options = [])
             $attributes[$attributeConf['id']] = $attributeConf;
         }
 
-        $rows = [];
+        $rows              = [];
```
Comment (BitOne) [inline src/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/AssociatedProductHydrator.php]: nop

File: src/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/AssociatedProductHydrator.php [hunk 6]
```diff
@@ -52,11 +53,44 @@ public function hydrate($queryBuilder, array $options = [])
...
-            $collection      = $documentManager->getDocumentCollection($this->productClass);
-            $results         = $collection->aggregate($pipeline)->toArray();
+            $associatedIds     = $this->getAssociatedProductIds($currentProduct, $associationTypeId);
+            $nbTotalAssociated = count($associatedIds);
+
+            $limit    = $queryDefinition['limit'];
+            $skip     = $queryDefinition['skip'];
+            $rawQuery = $queryDefinition['query'];
+
+            //first query: fetch only associated for this page
```
Comment (BitOne) [inline src/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/AssociatedProductHydrator.php]: Comment Paraphrasing  the code, pleasr remove.

File: src/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/AssociatedProductHydrator.php [hunk 7]
```diff
@@ -52,11 +53,44 @@ public function hydrate($queryBuilder, array $options = [])
...
+            $queryDefinition['skip'];
+            $associatedProducts = $this->getAssociatedProducts(
+                $queryBuilder,
+                $associatedIds,
+                $rawQuery,
+                $limit,
+                $skip
+            );
+
+            //second query: fetch only non associated for this page
```
Comment (BitOne) [inline src/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/AssociatedProductHydrator.php]: idem here.

File: src/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/AssociatedProductHydrator.php (hunk 2 above)
Comment (BitOne) [inline src/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/AssociatedProductHydrator.php]: Useless comment

File: src/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/AssociatedProductHydrator.php [hunk 8]
```diff
@@ -75,12 +109,12 @@ public function hydrate($queryBuilder, array $options = [])
         $assocTransformer  = new AssociationTransformer();
 
         foreach ($results as $result) {
-            $result = $fieldsTransformer->transform($result, $locale);
-            $result = $valuesTransformer->transform($result, $attributes, $locale, $scope);
-            $result = $familyTransformer->transform($result, $locale);
-            $result = $complTransformer->transform($result, $locale, $scope);
-            $result = $groupsTransformer->transform($result, $locale, $groupId);
-            $result = $assocTransformer->transform($result, $associationTypeId, $currentProduct);
+            $result               = $fieldsTransformer->transform($result, $locale);
```
Comment (BitOne) [inline src/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/AssociatedProductHydrator.php]: A good example of why alignement is a bad idea in this kind of context...

File: src/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/AssociatedProductHydrator.php [hunk 9]
```diff
@@ -52,11 +53,44 @@ public function hydrate($queryBuilder, array $options = [])
...
+            );
+
+            //second query: fetch only non associated for this page
+            $nonAssociatedProducts = [];
+            $nbAssociated = count($associatedProducts);
+            if ($limit > $nbAssociated) {
+                $limit -= $nbAssociated;
+                $skip -= $nbTotalAssociated;
+                // $skip can be < 0  for example when you have 12 associated products, each page contains 10 items,
+                // and you're looking for the second page
```
Comment (BitOne) [inline src/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/AssociatedProductHydrator.php]: To remove, useless.

File: src/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/AssociatedProductHydrator.php [hunk 10]
```diff
@@ -52,11 +53,39 @@ public function hydrate($queryBuilder, array $options = [])
...
+                $limit,
+                $skip
+            );
+
+            $nonAssociatedProducts = [];
+            $nbAssociated = count($associatedProducts);
+            if ($limit > $nbAssociated) {
+                $limit -= $nbAssociated;
+                $skip -= $nbTotalAssociated;
+                $skip = ($skip > 0) ? $skip : 0;
```
Comment (grena) [inline src/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/AssociatedProductHydrator.php]: oh oh oh, I think there is an error here

File: src/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/AssociatedProductHydrator.php [hunk 11]
```diff
@@ -90,56 +120,99 @@ public function hydrate($queryBuilder, array $options = [])
...
-
-        return $pipeline;
+        return $ids;
+    }
+
+    /**
+     * Get the associated products for the current page
+     *
+     * @param Builder $queryBuilder
+     * @param array   $associatedIds
```
Comment (jmleroux) [inline src/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/AssociatedProductHydrator.php]: `string[]`

File: src/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/AssociatedProductHydrator.php [hunk 12]
```diff
@@ -90,56 +120,99 @@ public function hydrate($queryBuilder, array $options = [])
...
+        );
+
+        return $results;
+    }
+
+    /**
+     * Get the non associated products for the current page
+     *
+     * @param Builder $queryBuilder
+     * @param array   $associatedIds
```
Comment (jmleroux) [inline src/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/AssociatedProductHydrator.php]: `string[]`

**Comment by jmleroux** [comment]
GTM

File: src/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/AssociatedProductHydrator.php (hunk 10 above)
Comment (jjanvier) [inline src/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/AssociatedProductHydrator.php]: learn how to read the code :trollface:

**Comment by jjanvier** [comment]
grena's comment is false

File: src/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/AssociatedProductHydrator.php [hunk 13]
```diff
@@ -90,56 +120,99 @@ public function hydrate($queryBuilder, array $options = [])
...
+        $in = ['_id' => ['$in' => $associatedIds]];
+        $rawQuery['$and'][] = $in;
+
+        $queryBuilder->setQueryArray($rawQuery);
+        $queryBuilder->limit($limit);
+        $queryBuilder->skip($skip);
+
+        $query   = $queryBuilder->getQuery();
+        $results = $query->execute()->toArray();
+        array_walk(
```
Comment (phaseinducer) [inline src/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/AssociatedProductHydrator.php]: As @BitOne said, foreach can do the same and is more readable.

File: src/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/AssociatedProductHydrator.php [hunk 14]
```diff
@@ -90,56 +120,99 @@ public function hydrate($queryBuilder, array $options = [])
...
+        $nin = ['_id' => ['$nin' => $associatedIds]];
+        $rawQuery['$and'][] = $nin;
+
+        $queryBuilder->setQueryArray($rawQuery);
+        $queryBuilder->limit($limit);
+        $queryBuilder->skip($skip);
+
+        $query   = $queryBuilder->getQuery();
+        $results = $query->execute()->toArray();
+        array_walk(
```
Comment (phaseinducer) [inline src/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/AssociatedProductHydrator.php]: same

File: src/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/AssociatedProductHydrator.php (hunk 13 above)
Comment (BitOne) [inline src/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/AssociatedProductHydrator.php]: Replace with foreach please.

File: src/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/AssociatedProductHydrator.php (hunk 14 above)
Comment (BitOne) [inline src/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/AssociatedProductHydrator.php]: Idem than above.

File: src/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/AssociatedProductHydrator.php [hunk 15]
```diff
@@ -90,56 +120,99 @@ public function hydrate($queryBuilder, array $options = [])
...
+        Builder $queryBuilder,
+        array $associatedIds,
+        array $rawQuery,
+        $limit,
+        $skip
+    ) {
+        $in = ['_id' => ['$in' => $associatedIds]];
+        $rawQuery['$and'][] = $in;
+
+        $queryBuilder->setQueryArray($rawQuery);
```
Comment (BitOne) [inline src/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/AssociatedProductHydrator.php]: Can you refactor this piece of code from there to the end to another function ? So we will have the code only once in the code ?

**Comment by BitOne** [comment]
GTM !

**Pull request merged**